### PR TITLE
Correct ghost atoms bug in fix adapt

### PR DESCRIPTION
**Pull request by jibril-b-coulibaly**
**Summary**

Dear developers,

This PR corrects a bug related to a recent implementation (by myself) of the `scale` keyword for parameters `diameter` and `scale` in #1937 (4acdd16). It also highlights an issue I could not pin down.

The `post_constructor()` function stores diameter and/or charge of ***local*** atoms in `fix_diam->vstore` and `fix_chg->vstore` respectively. However, the `change_settings()` function modifies the diameter/charge of ***both local and ghost*** atoms. When using the `scale` keyword, the `fix->vstore` array is therefore accessed for indices that are not defined, causing undefined behavior.

**Related Issues**

Issue #1508 already closed by previous PR #1937

**Author(s)**

Jibril B. Coulibaly, Northwestern University (jibril.coulibaly@northwestern.edu)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No change

**Implementation Notes**

This PR corrects this bug by storing the initial diameter/charge of both local and ghost atoms in `fix->vstore` in the `post_constructor()` function so that the array is always accessed where it is defined. The diameter/charge of ghost atoms is also restored in the `restore_settings()` function assuming it would prevent ghost atoms to potentially carry on the modified diameter/charge. Prior to 4acdd16, it seemingly caused no issue that both local and ghost atoms were modified in `change_settings()` but only local atoms were restored in `restore_settings()` so I am not exactly sure of the effectiveness of that second modification.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

The MWEs provided in #1937 did not trigger this issue. A more complex simulation for which I provide a restart file `restart.sample_preparation_3` and associated MWEs does.
- The input file `in.bug_ghost` illustrates the undefined behavior issue where values of pressure are altered with no actual change of diameter. No issue occurs with the proposed patch.
- The input file `in.bug_multiple_runs` shows the issue I could not resolve where consecutive runs result in abnormal pressure for 2 identical systems, unless a `run 0` is performed between these runs.
[MWE_fix_adapt_ghost_PR.zip](https://github.com/lammps/lammps/files/4648572/MWE_fix_adapt_ghost_PR.zip)

### CODE REVIEW

**Comment by akohlmey** [comment]
@jibril-b-coulibaly not sure if this approach will work. Not all property arrays may be available for local and ghost atoms, so copying and resetting them may lead to segmentation faults. Normally, only data for local atoms would need to be stored and restored, if you run a forward communication after the restore (or modification), which will update the per-atom data on ghost atoms supported by the atom style from their local originals to their ghost copies. Also you can only propagate local data with atoms during a run and then restore it at the end. ghost atoms will vanish or be created new.

**Comment by jibril-b-coulibaly** [comment]
Thank you very much for your insightful analysis @akohlmey. I am reaching the edges of my understanding.

I gather that if only local atoms should have their values changed/restored, a different approach must be used to alter the diameter of ghost atoms.

To perform the force calculation correctly, there would be a need to alter the diameter of ghost atoms in `change_settings()`, as was done prior to 4acdd16 when using `scale no`. This is what prompted me to store them.

That appears to be more complex than I initially envisionned in #1937, which might therefore have to be reverted or amended, and which might explain why the `scale` keyword was not implemented for charge and diameter in the first place.

**Comment by jibril-b-coulibaly** [comment]
My understanding is that in order for the scaling of the diameter to work, the per-atom data of ghost atoms should be updated every step. This can however only be done by communicating the local data (which can be safely modified/restored without running into memory issues) to the ghost atoms at every step. This would likely explain the issue with the multiple runs I encountered.

I just found out about the `radvary` variable in [AtomVec class](https://github.com/lammps/lammps/blob/master/src/atom_vec_sphere.cpp) that appears to control communications when `fix adapt` is used but is not yet documented in the manual. I did not have that commit in my repo and it appears recent and important refactoring of `AtomVec` class occurred in #1788. The latest version (b1ee9ad1f) now makes my previous MWEs crash with:
`ERROR: Fix adapt changes particle radii but atom_style sphere is not dynamic (C:\Users\jibri\Desktop\lammps\src\atom_vec_sphere.cpp:98)`
I converted my restart file into a data file and used `atom_style sphere 1` in a new enclosed [MWE](https://github.com/lammps/lammps/files/4653208/MWE_fix_adapt_data.zip).

These changes however do not solve the present issue, seemingly indicating that the ghost atoms are not being given the current radius of local atoms at every timestep. To illustrate this:

- with the original code, memory issues from dereferencing `vec[i]` for `i>nlocal` occur (see bogus values of pressure in `log_original.lammps`)
- with the proposed PR, the results work as expected (see `log_nall_everywhere.lammps`) but we agree that this is not a viable solution and could cause segmentation faults
- I tried altering the diameter of the local atoms only by changing `nall` to `nlocal` in `change_settings()` line 586-588 :
```diff
- int nall = nlocal + atom->nghost;
- for (i = 0; i < nall; i++)
+ for (i = 0; i < nlocal; i++)
```
to see if such local modification would be communicated to ghost atoms, without success. The pressure is smaller compared to when the diameters are modified to an identical value using `set diameter` (see `log_nlocal_change_settings.lammps`), seemingly indicating smaller diameters in ghost atoms.

I am kindly seeking your guidance on how to best solve this issue. Facing the impossibility to safely store ghost atom diameters, my intuition to scale the diameter would be to either:


1. communicate the initial diameter of the ghost atoms from their original local counterparts but I do not know if this is feasible and if the corresponding mapping/callbacks exist to do so without significant overhead.
2. store the previous value `value_prev` of the variable every `N` timestep and modify the current radius as `radius[i] = radius[i]*value/value_prev` for `nall` atoms thereby not using the initial values of the `fix store`. I will start implementing that solution if you consider no simple or efficient solution is achievable through communication of the `fix_store` data.

Thank you very much

**Comment by sjplimp** [comment]
@jibril-b-coulibaly @akohlmey ok - there are a few things to clarify here

(a) when we approved the earlier PR that added scale as an option to diameter we should have caught the issue that this it didn't work for ghost atoms - that is our (reviewers) fault

(b) scale for diameter requires storing the original diameter with each atom via fix STORE,
which you are already doing.  Fix STORE does nothing for ghost atoms which is correct behavior

(c) I presume that before your previous PR, that change_settings() for diameter was looping over owned and ghost atoms and resetting the diameter

(d) all that you should need to add to that (for diameter and charge) in change_settings() is a communication call (only if scale is enabled) that performs forward comm on the vector stored by fix STORE.  This requires you add a pack and unpack method to fix adapt.  These are functions called back to by comm->forward_comm_fix().  This will make sure ghost atoms have a current copy of the original diameter of the particle.  Then the loop over nall = nlocal+nghost to reset diameters should work.

(e) The logic in AtomVecSphere with radvary is to insure that radius (and rmass) is communicated on all other steps, not just the step when fix adapt changes them.  Now that I'm thinking about this more carefully, I'm not sure it's actually necessary.  Once you have fix adapt working as you expect (with or without scale) you could test if disabling that part of atom_vec_sphere makes any difference.  I'm now thinking it won't.

(f) The use of the 0/1 flag for radvary for spherical particles is documented in the atom_style command, with the AtomVec refactor PR - it isn't in the website doc pages b/c that patch release hasn't appeared yet.

@akohlmey Are there any other options for changing particle diameters/mass in between reneighborings that would require the radvary logic in AtomVecSphere?  I can't think of any besides fix adapt.  Which I think will not need it, since it sets ghost atom diameter/mass directly.

**Comment by sjplimp** [comment]
@jibril-b-coulibaly Also, if these granular particles are part of rigid bodies, then
changing their diameter/mass via fix adapt will alter the body moments of inertia, which I don't think is being accounted for?

**Comment by jibril-b-coulibaly** [comment]
Thank you very much @sjplimp for a detailed point by point answer.

(c) yes, prior to #1937, `change_settings()` was looping over the diameter of both local and ghost atoms, affecting `radius[i] = 0.5*value`.

(e) Thank you for the clarifications. I believe this should not make a difference as well since the values of diameter do not change during the concerned steps.

(2nd comment) Changes in rigid molecules made of granular atoms are not (never were) accounted for and would require rebuilding the rigid molecules. In terms oh physics/geometry however, changing only the diameter would not correspond to a proper change in dimensions of the rigid molecule. It could change the topology, e.g. holes appearing for decreasing diameters. The relative distance between the particles would have to change by the same factor to produce a proper change of dimensions. We could therefore warn about the use of the fix for granular particles belonging to rigid molecules, or restrict its use at the moment.

As suggested in my previous comment, I implemented another method that incrementally performs the scaling, thereby avoiding the use of the stored data in the computation altogether. This seems to work with the MWEs and does not require communication.

##

There remains one aspect that is unclear to me about (d) and I would be very grateful if you would be willing to discuss it.

From (b), I understand that fix STORE stores the initial diameter of the local atoms only, 
so that each proc has a fix whose `fix_diam->vstore` vector contains nlocal values of initial diameter.

From (d), for `comm->forward_comm_fix()` to communicate ghost atoms their initial diameter, I need to pack the nlocal values of initial diameter `fix_diam->vstore` of a given proc using `fix_adapt::pack_forward_comm` and then unpack these values into a different proc using `fix_adapt::unpack_forward_comm`. My understanding is that they will be unpacked into the ghost atoms of that proc (`i>nlocal`) during the forward communication.

That seems to bring me back to the original issue of `fix_diam->vstore` holding information for ghost atoms `i>nlocal` and to contradict (b). 

>  This will make sure ghost atoms have a current copy of the original diameter of the particle.

It is unclear to me, in terms of memory/faults, how communicating the initial diameter from the original atom that is local to a different proc, and storing it into the `fix_diam->vstore`, differs from storing the initial diameter of the ghost atom directly into `fix_diam->vstore`.
Are the memory/segfault issues stemming from the fact that [fix STORE](https://github.com/lammps/lammps/blob/master/src/fix_store.h) does not implement the `pack_forward_comm` and `unpack_forward_comm` methods ?



Thank you

**Comment by sjplimp** [comment]
@jibril-b-coulibaly The fix STORE vector has space to store ghost atom values.  But it does not do any communication for those values (nor should it).  I'm suggesting you trigger the necessary comm in fix adapt and populate the ghost values in that vector so the loop to nall = nlocal + nghost works.  Once that loop is done, you should assume those ghost values are no longer there - they will be wiped out the next time neighbor lists are rebuilt.

Re: rigid bodies - you can look at the "molecule scale" command which explains
how scaling is done for granular molecules (e.g. for rigid bodies).  This is also used
by fix pour (e.g. to pour rigid bodies into a simulation box) and "fix rigid/small mol".
In principle something like this could be made to work with fix adapt to alter
granular rigid bodies over time.

**Comment by jibril-b-coulibaly** [comment]
Thank you very much @sjplimp and @akohlmey. I think I am finally piecing your comments together and understanding why my proposal does not work:

> I'm suggesting you trigger the necessary comm in fix adapt and populate the ghost values in that vector [...]. Once that loop is done [...] those ghost values will be wiped out the next time neighbor lists are rebuilt.

My error in e1362e9 was that the ghost values were populating the vector in `post_constructor()`, which is only called once. As soon as the neighbor list was rebuilt, these values were wiped out. Therefore looping over nall=nlocal+nghost in `change_settings()` and `restore_settings()` dereferenced the vector ouf of bounds.

Populating the vector in `change_settings()` would resolve the issue. That being said, I would like to know your opinion on 03419b3.

Instead of communicating and scaling the initial diameter by the current value of the fix adapt variable:
_**diam[N] = diam[0] * value[N]**_
this solution implements a memory variable whose initial value is 1 and that is updated each step fix adapt changes the value. The diameter is updated at each step as:
 _**diam[N] = diam[N-1]*value[N]/value[N-1]**_
By recurrence, it gives the same relationship diam[N] = diam[0] * value[N], and the same numerical value (to within rounding precision?).

It is not as "canonical"  but appears to perform scaling correctly, and potentially for a lower cost/overhead than the forward communication.

**Comment by sjplimp** [comment]
@jibril-b-coulibaly The recurrence relation is a clever way to do it.
Just 3 things to consider:

1) Will it be accurate if someone invokes fix adapt every step for a billion
steps.  Someone will do that at some point.

2) Fix adapt also has to allow for restarting a long simulation.
So you would need to make sure you enable the ability
to use the previous step scale value even when it is the
first step of a restarted simulation.  Either store the
previous value in the restart file or recompute it In
the new run.

3) Does your continually changing the values in the fix STORE
vector work with the fix adapt reset option?  That was
the original motivation for using fix STORE.  So that the
original radii or charges would be available after the run
to reset them.

**Comment by jibril-b-coulibaly** [comment]
Thank you very much for these suggestions.

(1) I implemented both the recurrence (03419b3) and the forward communication (https://github.com/jibril-b-coulibaly/lammps/tree/scale_comm) for comparison. I have scaled values of a single atom every step for 10 million steps and get identical results for recurrence and forward communications (see in.scale_accuracy enclosed). I also carried out tests on 8 procs, applying fix adapt every step, to assess the performances of each solution.
- For a single node (Intel Haswell E5-2680), I did not notice any significant difference
- For multiple nodes (Intel Haswell E5-2680 + Infiniband QDR), the forward communication is significantly slower than the recurrence, with increased time spent in "Modify".
When applying fix adap every 10 step, loop times become comparable (see in.scale_cost and logs attached).

[MWEs_accuracy_cost.zip](https://github.com/lammps/lammps/files/4669367/MWEs_accuracy_cost.zip)

(3) For the recurrence, values in fix store are never used or changed, initial values can therefore be reset properly without issue.

(2) Regarding saving the current scaling factor in the restart file, I am not sure if that is necessary/practical. If we save the current value of the time-dependent variable used to scale properties, the user would also have to restart it at the exact same value, for the simulation to resume correctly. This could be a complication if the user defined the variable using `ramp` for example, or if they wanted a different scaling in a new restarted simulation. 
On a side note, I was trying to implement a switch to get the value from the restart file if the fix is restarted or to set it to 1 if not (similar to what is done for the fix STORE in `post_constructor()`). I noticed the flag `restart_reset` is not assigned to 1 for Fixes in restart global list, in [modify.cpp](https://github.com/lammps/lammps/blob/master/src/modify.cpp) (lines 909-948). Is this normal? Would assigning value 1 cause other issues ? Or should I used the `used_restart_global` flag instead?

Thank you

**Comment by sjplimp** [comment]
@jibril-b-coulibaly Re: (2), all I am saying is that users will want
to use fix adapt for long simulations where they restart multiple times.
If you are saying they can do that correctly by carefully choosing
the parameters for their variable or ramp or fix adapt in the restart
script, then that is OK.  But the doc page should probably explain it.

Since your recurrence formula uses value[n-1], I assumed you might
want/need to know that value when you restart, to make it easier
to restart.

Also note that as explained on the variable doc page, the ramp function
uses the optional run command settings to allow multiple runs that
span a larger number of time steps.  So that the value returned by
ramp() can span many runs.  That's one way users may want to
perform multiple restarts, and have their fix adapt values increase
smoothly across all the runs.

**Comment by sjplimp** [comment]
@jibril-b-coulibaly Re: your Q about restart flags,
not sure what you are asking.  If you wanted fix adapt
to store a global value (like value[n-1]), then
you set restart_global = 1 in the constructor,
and add write_restart() and restart() methods.
Do this grep to see examples:
grep restart_global fix*cpp

Steve

**Comment by jibril-b-coulibaly** [comment]
@sjplimp I implemented the corresponding functions in 1add74b.
I asked for the restart flag because I would like to default the scale factor to 1 in `init()` when the value is neither read from the restart file, nor the continuation of a previous run. The expected behavior for a fresh run should be that the values are scaled exactly according to the scaling factor, which requires the scaling factor to be set to 1 at the beginning of each fresh run.

The following example illustrates it:
```
units si
dimension 3
boundary p p p
atom_style sphere 1
comm_modify	vel yes
newton off
atom_modify map yes
neigh_modify delay 0
neighbor 0.1 bin

region box block -2 2 -2 2 -2 2
create_box 1 box
create_atoms 1 single 0 0 0
set atom 1 diameter 1.0

compute 1 all property/atom diameter
variable diam equal c_1[1]

################
# Fresh run, ramp diameter from 1 to 3
variable scale_factor equal "ramp(1,3)"
fix 1 all adapt 1 atom diameter v_scale_factor scale yes reset yes
restart 50 restart.scale*
thermo 50
run 100

clear

################
# Restarted run
compute 1 all property/atom diameter
variable diam equal c_1[1]

read_restart restart.scale50
variable scale_factor equal "ramp(2,3)" # Dtopped at scaling of 2, want to restart at 2
fix 1 all adapt 1 atom diameter v_scale_factor scale yes reset yes
thermo_style custom step v_diam
thermo 10
run 50

################
# Fresh run again
variable scale_factor equal "ramp(1,3)"
run 0 # This is a fresh run again, expected diameter is 1
```

- With 1add74b, and the `diam_scale=1` hardcoded in the `init()` function, the scale factor saved in the restart file is overwritten. Therefore, we have the wrong behavior during the second (restart) run starting with diameter=4 instead of diameter=2, but the correct one during the third (fresh) run with diameter=1.
- With 6d11515, and the `diam_scale=1` the default in the constructor, the scale factor saved in the restart file is applied correctly during the restart. Therefore, we have the correct behavior during the second (restart) run starting with diameter=2. However, the last value in memory of the scale factor is 3, so when the third (fresh) run occurs, we get the wrong value diameter=1/3.

Both commits have issues that can be solved with a switch in the `init()` function that would look something like `if(!restart_reset) diam_scale = 1;`. The `restart_reset` flag however is not defined for fixes that store global quantities so I could not use it, hence my questions about how to proceed.

Thank you

**Comment by jibril-b-coulibaly** [comment]
Grep shows the `restart_reset` flag in (besides fix.h and fix.cpp):

- `compute_displace_atom.cpp`, `compute_msd.cpp` and `compute_vacf.cpp` (a Fix Store stores per-atom arrays of initial position or velocity)
- `fix_adapt.cpp` (a Fix store stores per-atom vectors of initial diameter and charge)
- `modify.cpp` (flag set to 1 when restarting fixes storing per-atom information)

The flag is currently only defined, and used, for fixes in the restart_peratom list ([modify.cpp](https://github.com/lammps/lammps/blob/5e3fe19756d6943975790759fd9781d462ee72f6/src/modify.cpp#L938)). I believe it should also be set for fixes in the restart_global list.

a6d2ae2 implements this, and defaulting values of scale parameters to 1 if the fix is not restarted, providing the expected behavior upon restart and upon fresh runs. This should complete the proper implementation/debugging of the scale keywords for diameter and charge.

**Comment by sjplimp** [comment]
@jibril-b-coulibaly I just pushed some modest changes to this PR.  You should both run it on your test cases to verify it still works as you expect.  And also read the fix adapt doc page to see if you agree with the added info about restarting with this fix.  Aside from cosmetic edits and clarifying how restarts can work, my chief code changes were:
-- don't write global restart info unless atom and scale are set, also insure the two scale values are initialized, otherwise were sometimes writing/reading uninit values to the restart file I believe.

**Comment by jibril-b-coulibaly** [comment]
@sjplimp thank you for thorough a review and useful changes. I reviewed them, tested my MWEs and everything works as expected. The doc information is clear, thank you for that addition.

I performed additional modifications based on the following observations:

**1. Numerical accuracy for mass changes**
*Issue*:
In some cases, I noticed divergence in dynamic runs when using either `reset yes` or `reset no`, and when the dynamic run was preceded by a `run 0` (see `div_press_reset` files in attachment). I found the issues to stem from machine precision in the way the `rmass` is currently modified (see `rmass_reset` files in attachment). 
Diameter changes, even made with a scale = 1, can result in minute differences of mass (see `dmp_rmass_reset_no`, `dmp_rmass_reset_yes`). This causes the divergence observed in dynamic runs: a tiny change in mass occurs during the `run 0` (see `dmp_rmass_reset_no`, `dmp_rmass_reset_yes`), and subsequent dynamic runs diverge. This is further confirmed by the fact that no divergence is observed if the `run 0` is removed or placed before the definition of the `fix adapt`.
*Solution*:
This issue is fixed by b923037, scaling the current mass by (diam[N]/diam[N-1])^dim, where dim=2,3 is the simulation dimensions. This appears to limit rounding errors and provides consistent results regardless of the `reset` keyword and precedence of the `run 0`.

**2. Optional mass change**
*Issue*:
I would like to perform diameter changes without mass changes. My motivation in this PR is to model thermal expansion of granular materials (e.g., Lu et al., 2001 https://doi.org/10.1016/S0022-3115(01)00690-0). In that case, thermal expansion changes the diameter but does not cause any change in mass of the particles.
*Solution*: 
This issue is fixed by 47ebd8a3d, implementing a `mass` keyword and `massflag` that determines whether or not the mass is changed when the diameter is changed. The documentation also mentions that possibility.
[MWEs_divergence_rmass_reset.zip](https://github.com/lammps/lammps/files/4821147/MWEs_divergence_rmass_reset.zip)

**Comment by stanmoore1** [comment]
Allowing the mass to change may break other things. For example, I believe the KOKKOS package assumes the mass is constant and only copies the value once from CPU to GPU.

**Comment by stanmoore1** [comment]
I took a look at the changes and I think this is actually OK wrt KOKKOS. I though you were changing `atom->mass` , as opposed to `atom->rmass`.

File: doc/src/fix_adapt.rst [hunk 1]
```diff
@@ -107,16 +111,17 @@ style supports it.  Note that the :doc:`pair_style <pair_style>` and
...
-second, etc.  For example, *pstyle* could be specified as "soft" or
-"lubricate" or "lj/cut:1" or "lj/cut:2".  The *pparam* argument is the
-name of the parameter to change.  This is the current list of pair
-styles and parameters that can be varied by this fix.  See the doc
-pages for individual pair styles and their energy formulas for the
-meaning of these parameters:
+The *pstyle* argument is the name of the pair style.  If
+:doc:`pair_style hybrid or hybrid/overlay <pair_hybrid>` is used,
+*pstyle* should be a sub-style name.  If there are multiple
+sub-stylesusing the same pair style, then *pstyle* should be specified
```
Comment (sjplimp) [inline doc/src/fix_adapt.rst:117]: needs a space after sub-styles

File: doc/src/fix_adapt.rst [hunk 2]
```diff
@@ -45,6 +45,9 @@ Syntax
        *reset* value = *no* or *yes*
          *no* = values will remain altered at the end of a run
          *yes* = reset altered values to their original values at the end of a run
+       *mass* value = *no* or *yes*
+         *no* = mass is not altered by changes in diameter
+         *yes* = mass is altenred by changes in diameter
```
Comment (sjplimp) [inline doc/src/fix_adapt.rst:50]: typo in altered

**Comment by sjplimp** [comment]
@jibril All looks good now.  Just fix the couple of typos and this should be ready to merge.
Thanks for all your careful work and testing on this.

**Comment by akohlmey** [comment]
Updated to current master and fixed a couple more spelling errors. Marking ready for review now.

**Pull request merged**